### PR TITLE
 Add post listeners to all get authenticator service methods

### PR DESCRIPTION
**Pull request by senthalan**
### Proposed changes in this pull request

Provide the capability to perform some post actions on the returned list of authenticators, identity provider and request path authenticator

### CODE REVIEW

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationManagementServiceImpl.java [hunk 1]
```diff
@@ -780,60 +780,93 @@ public IdentityProvider getIdentityProvider(String federatedIdPName, String tena
     public IdentityProvider[] getAllIdentityProviders(String tenantDomain)
             throws IdentityApplicationManagementException {
 
+        List<IdentityProvider> fedIdpList;
```
Comment (mefarazath) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationManagementServiceImpl.java:783]: I know we are currently using this method (in ApplicationManagementServiceImpl.java ) to get metadata related to IDPs, local authenticators, request path authenticators etc.

But is ApplicationManagementServiceImpl.java the right place to get this info?

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationManagementServiceImpl.java [hunk 2]
```diff
@@ -780,60 +780,93 @@ public IdentityProvider getIdentityProvider(String federatedIdPName, String tena
...
+            fedIdpList = idpdao.getAllIdentityProviders();
         } catch (Exception e) {
             String error = "Error occurred while retrieving all Identity Providers" + ". " + e.getMessage();
             throw new IdentityApplicationManagementException(error, e);
         } finally {
             endTenantFlow();
         }
+
+        // Invoking the listeners
+        Collection<ApplicationMgtListener> listeners =
```
Comment (Thumimku) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationManagementServiceImpl.java:796]: Don't we need to check fedIdp nullability before invoking listener doPostGetAllIdentityProviders? Can we pass a null value to this method?

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/listener/ApplicationMgtListener.java [hunk 3]
```diff
@@ -387,4 +392,36 @@ default boolean doPreUpdateApplicationTemplate(ServiceProvider serviceProvider,
...
         return true;
     }
+
+    /**
+     * Define any additional actions to the get all local authenticator method.
+     *
+     * @param tenantDomain The requested tenant domain.
+     * @param localAuthenticators   The local authenticators to be returned by the service.
+     */
+    default void doPostGetAllLocalAuthenticators(String tenantDomain,
```
Comment (mefarazath) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/listener/ApplicationMgtListener.java:402]: This listener should only be related to actions related to the application.
Filtering metadata using listeners doesn't look right.

Comment (IsuraD) [reply components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/listener/ApplicationMgtListener.java:402]: what is the purpose of introducing these listeners?

Comment (senthalan) [reply components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/listener/ApplicationMgtListener.java:402]: This is to give extendability to do perform any operation on the returned local authenticators. The main use case will be to filter out any authenticators per tenant.

**Pull request closed without merging**